core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/List2OptMove.java

```diff
@@ -0,0 +1,231 @@
+    }
+
+    @Override
+    protected List2OptMove<Solution_> createUndoMove(ScoreDirector<Solution_> scoreDirector) {
+        return new List2OptMove<>(variableDescriptor,
+                                  indexVariableSupply,
+                                  entity,
+                                  firstEdgeStartpoint,
+                                  secondEdgeStartpoint,
+                                  firstEdgeEndpoint,
+                                  secondEdgeEndpoint);
+    }
+
+    @Override
+    protected void doMoveOnGenuineVariables(ScoreDirector<Solution_> scoreDirector) {
+        InnerScoreDirector<Solution_, ?> innerScoreDirector = (InnerScoreDirector<Solution_, ?>) scoreDirector;
+        List<Object> listVariable = variableDescriptor.getListVariable(entity);
+        int firstEdgeEndpointIndex = indexVariableSupply.getIndex(firstEdgeEndpoint);
+        int secondEdgeEndpointIndex = indexVariableSupply.getIndex(secondEdgeEndpoint);
+
+        if (firstEdgeEndpointIndex < secondEdgeEndpointIndex) {
+            innerScoreDirector.beforeListVariableChanged(variableDescriptor, entity,
+                                                         firstEdgeEndpointIndex,
+                                                         secondEdgeEndpointIndex);
+
+            Collections.reverse(listVariable.subList(firstEdgeEndpointIndex, secondEdgeEndpointIndex));
+
+            innerScoreDirector.afterListVariableChanged(variableDescriptor, entity,
+                                                        firstEdgeEndpointIndex,
+                                                        secondEdgeEndpointIndex);
+        } else {
+            List<Object> firstHalfReversedPath = listVariable.subList(firstEdgeEndpointIndex, listVariable.size());
+            List<Object> secondHalfReversedPath = listVariable.subList(0, secondEdgeEndpointIndex);
+
+            innerScoreDirector.beforeListVariableChanged(variableDescriptor, entity,
+                                                         firstEdgeEndpointIndex,
+                                                         listVariable.size());
+            innerScoreDirector.beforeListVariableChanged(variableDescriptor, entity,
+                                                         0,
+                                                         secondEdgeEndpointIndex);
+
+            // Reverse the combined list firstHalfReversedPath + secondHalfReversedPath
+            // For instance, (1, 2, 3)(4, 5, 6, 7, 8, 9) becomes
+            // (9, 8, 7)(6, 5, 4, 3, 2, 1)
+            int totalLength = firstHalfReversedPath.size() + secondHalfReversedPath.size();
+            for (int i = 0; (i < totalLength >> 1); i++) {
+                if (i < firstHalfReversedPath.size()) {
+                    if (i < secondHalfReversedPath.size()) {
+                        // firstHalfIndex = i
+                        int secondHalfIndex = secondHalfReversedPath.size() - i - 1;
```

Comment:
I suggest splitting off the branches into individual private methods; that way, they will be more easily benchmarked. Flame graphs will show the branch, and not the `doMove...()` method as a whole.